File: src/main/scala/algo/WordMachine.java
package algo;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Stack;
import java.util.function.Supplier;
import java.util.regex.Pattern;

public class WordMachine {
    public static final int MIN = 0;
    public static final int MAX = 0x7FFF_FFFF;
    private static final Pattern NUMERIC_PATTERN = Pattern.compile("\\d+");
    private final Stack<Integer> stack = new Stack<>();
    private static final Map<String, Supplier<Integer>> FUNCTIONS = new HashMap<>();
    {
        FUNCTIONS.put("POP", this::pop);
        FUNCTIONS.put("DUP", this::dup);
        FUNCTIONS.put("+", this::add);
        FUNCTIONS.put("-", this::sub);
    }

    public static void main(String[] args) {
        WordMachine wm = new WordMachine();
        System.out.println(wm.process("1 1 + 2 + DUP +"));
    }

    public int process(String s) {
        try {
            Arrays.stream(s.split(" ")).forEach(this::apply);
            return pop();
        } catch (IllegalArgumentException e) {
            return -1;
        }
    }

    private void apply(String s) {
        if (NUMERIC_PATTERN.matcher(s).matches()) {
            push(Integer.valueOf(s));
        } else {
            FUNCTIONS.get(s).get();
        }
    }

    private int push(int i) {
        withinRange(i);
        return stack.push(i);
    }

    private int pop() {
        return stack.pop();
    }

    private int dup() {
        hasElements(1);
        return push(stack.peek());
    }

    private int add() {
        hasElements(2);
        return push(stack.pop() + stack.pop());
    }

    private int sub() {
        hasElements(2);
        return push(stack.pop() - stack.pop());
    }

    private int hasElements(int i) {
        if(stack.size() < i){
            throw new IllegalArgumentException("Too few elements are available");
        }
        return i;
    }

    private int withinRange(int i){
        if(i < MIN || i > MAX){
            throw new IllegalArgumentException("Number out of integer range");
        }
        return i;
    }
}

File: src/main/scala/probls/simple/WordMachine.scala
package probls.simple

object WordMachine extends App {

  private val MIN = 0
  private val MAX = 0x7FFF_FFFF
  private val NumericPattern = "\\d+".r

  type Stack = List[Int]
  type Result[T] = Either[String, T]

  val userInput = "1 1 + 2 + DUP +"
  println(process(userInput))

  def process(input: String): Result[Int] =
    input
      .split(" ")
      .toList
      .foldLeft[Result[Stack]](Right(List.empty[Int])) {
        (stack, token) => apply(stack, token)
      }.map(_.head)

  private def apply(stack: Result[Stack], token: String): Result[Stack] = token match {
    case NumericPattern() => stack.flatMap(stack => push(stack, token.toInt))
    case "POP" => stack.flatMap(stack => pop(stack))
    case "DUP" => stack.flatMap(stack => dup(stack))
    case "+" => stack.flatMap(stack => add(stack))
    case "-" => stack.flatMap(stack => sub(stack))
    case _ => Left(s"Unknown operation: $token")
  }

  private def push(state: Stack, value: Int): Result[Stack] =
    withinRange(value).map(_.head :: state)

  private def pop(stack: Stack): Result[Stack] =
    stack match {
      case _ :: tail => Right(tail)
      case Nil => Left("Too few elements are available")
    }

  private def dup(stack: Stack): Result[Stack] =
    stack match {
      case head :: tail => Right(head :: head :: tail)
      case Nil => Left("Too few elements are available")
    }

  private def add(stack: Stack): Result[Stack] =
    stack match {
      case a :: b :: _ => Right(stack).flatMap(pop).flatMap(pop).flatMap(push(_: Stack, a + b))
      case _ => Left("Too few elements are available")
    }

  private def sub(stack: Stack): Result[Stack] =
    stack match {
      case a :: b :: _ => Right(stack).flatMap(pop).flatMap(pop).flatMap(push(_: Stack, a - b))
      case _ => Left("Too few elements are available")
    }

  private def withinRange(value: Int): Result[Stack] = Either
    .cond(value >= MIN && value <= MAX, List(value), "Number is out of integer range")

}
